### ISSUE TITLE
Separate arguments for query and key

### ISSUE THREAD

**Issue by Daniel63656** [SHI-Labs/NATTEN#61]
I am using NeighborhoodAttention2D but it only takes one argument during the forward pass. I want to use it in  decoder so I need to pass different key and query. These would have same dimensions in my case. I modified the function like this:

``
class NeighborhoodAttention2D(nn.Module):
    """
    Neighborhood Attention 2D Module
    """

    def __init__(
        self,
        dim,
        num_heads,
        kernel_size,
        dilation=1,
        bias=True,
        qkv_bias=True,
        qk_scale=None,
        attn_drop=0.0,
        proj_drop=0.0,
    ):
        super().__init__()
        self.num_heads = num_heads
        self.head_dim = dim // self.num_heads
        self.scale = qk_scale or self.head_dim**-0.5
        assert (
            kernel_size > 1 and kernel_size % 2 == 1
        ), f"Kernel size must be an odd number greater than 1, got {kernel_size}."
        self.kernel_size = kernel_size
        assert (
            dilation is None or dilation >= 1
        ), f"Dilation must be greater than or equal to 1, got {dilation}."
        self.dilation = dilation or 1
        self.window_size = self.kernel_size * self.dilation

        self.W_k = nn.Linear(dim, dim, bias=qkv_bias)
        self.W_v = nn.Linear(dim, dim, bias=qkv_bias)
        self.W_q = nn.Linear(dim, dim, bias=qkv_bias)

        if bias:
            self.rpb = nn.Parameter(
                torch.zeros(num_heads, (2 * kernel_size - 1), (2 * kernel_size - 1))
            )
            trunc_normal_(self.rpb, std=0.02, mean=0.0, a=-2.0, b=2.0)
        else:
            self.register_parameter("rpb", None)
        self.attn_drop = nn.Dropout(attn_drop)
        self.proj = nn.Linear(dim, dim)
        self.proj_drop = nn.Dropout(proj_drop)

    def forward(self, query, key_value):
        B, Hp, Wp, C = query.shape  # (B, H, W, d_m)
        H, W = int(Hp), int(Wp)
        pad_l = pad_t = pad_r = pad_b = 0
        if H < self.window_size or W < self.window_size:
            pad_l = pad_t = 0
            pad_r = max(0, self.window_size - W)
            pad_b = max(0, self.window_size - H)
            x = pad(x, (0, 0, pad_l, pad_r, pad_t, pad_b))
            _, H, W, _ = x.shape

        # create queries, keys and values
        q = self.W_q(query).    view(B, H, W, self.num_heads, self.head_dim).permute(0,3,1,2,4)
        k = self.W_k(key_value).view(B, H, W, self.num_heads, self.head_dim).permute(0,3,1,2,4)
        v = self.W_v(key_value).view(B, H, W, self.num_heads, self.head_dim).permute(0,3,1,2,4)

        q = q * self.scale
        attn = natten2dqkrpb(q, k, self.rpb, self.kernel_size, self.dilation)
        attn = attn.softmax(dim=-1)
        attn = self.attn_drop(attn)
        x = natten2dav(attn, v, self.kernel_size, self.dilation)
        x = x.permute(0, 2, 3, 1, 4).reshape(B, H, W, C)
        if pad_r or pad_b:
            x = x[:, :Hp, :Wp, :]

        return self.proj_drop(self.proj(x))

    def extra_repr(self) -> str:
        return (
            f"head_dim={self.head_dim}, num_heads={self.num_heads}, "
            + f"kernel_size={self.kernel_size}, dilation={self.dilation}, "
            + f"rel_pos_bias={self.rpb is not None}"
        )
``

**Comment by alihassanijr** [comment]
It is not advisable to use neighborhood attention in such cases, because neighborhood attention is only defined for self attention, and not for cross attention.

That said, if you can guarantee that the spatial structure between your query set and key-value pair is identical, you can use neighborhood attention, and your implementation (as far as I can tell) is correct.

**Comment by Daniel63656** [comment]
Actually, I want to use local attention in a "sparse" manner. Imagine you have an image partitioned into grid cells and only a few of these grids hold objects. I would like to have cross-attention, where only these selected grids attend to their local neighborhood respectively. This poses the same problem that is solved by natten, namely to select an attention window that is relative for each token.  The main difference would be that relative positions would not be as regular. 
Is there any way to include this? I would hapily contribute such a feature as I desperately need it. Maybe you could explain where to start/useful functions for this in more detail to me?

**Comment by alihassanijr** [comment]
Closing this; feel free to reopen if you have more questions.

**Comment by aniket-gupta1** [comment]
@Daniel63656 Were you able to make the Cross-Attention implementation work? If yes, does it come with any extra latency or does it run as fast as the self-attention version? 

I need CA for a project and I am thinking about doing the same thing, so your input will really help! Thanks in advance.

**Comment by alihassanijr** [comment]
@aniket-gupta1 is this what you're trying to do?
https://github.com/SHI-Labs/NATTEN/issues/82

If yes, we've already added the feature. You'd just need to pass `additional_keys` and `additional_values` to `na*d_qk` and `na*d_av` respectively (both imported from `natten.functional`.)

**Comment by aniket-gupta1** [comment]
@alihassanijr Thanks for your quick response! 
I just want to use Cross-attention with NA layers, so basically I have query from Image_1 features and key vector from Image_2 features. 

Can you please tell me how can do that?

**Comment by alihassanijr** [comment]
So that's kind of tricky. If you're assuming that the two images are the same resolution, so you end up with the same feature map size, you don't really need to do anything, and the standard NATTEN interfaces should work.
However, I can't really speak to whether or not it's a good idea, or its potential performance on your specific task.

If the two feature maps are shaped differently, I'm afraid there's not a straightforward way you can cross-attend the two, because you would need to 1. define a mapping from the query coordinate space to the KV coordinate space, and 2. implement NA kernels with that mapping.

Like all operators in deep learning libraries, NA operators are written under certain assumptions, the most critical of which is that we're reducing a self attention problem, therefore dealing with the same query and context coordinate space, in which the mapping is identity. Going beyond that requires a well defined mapping and the implementation to go with it.

**Comment by aniket-gupta1** [comment]
So, in my case the size of feature map is same for both images. To provide more clarity on what I want to do is, [GMFlow](https://github.com/haofeixu/gmflow/tree/main) is a optical flow paper which uses Global Attention layers. I want to check how well NA works in this case. 

So, in summary I have two feature maps of size (B,C,H,W) each and I want to apply cross-attention between them

**Comment by alihassanijr** [comment]
Yeah so in theory the only difference between self and cross attention is that the former assumes the query and context sets are the same (Q and KV are projections of the same input), whereas the latter assumes that they're not.

As far as implementation is concerned, as long as Q and KV have the same shape, you can use any of the ops/modules provided by NATTEN. (V technically can assume a different dimension per head, but it's a feature we'll be adding very soon.)

So based on what you've shared, NATTEN shouldn't get in the way of your use case, but whether it works or not depends on whether you can safely assume that there's an identity map between all feature maps you cross-attend together.

**Comment by aniket-gupta1** [comment]
Okay thanks for the clarification, my last doubt is what module can I use from Natten to directly implement cross-attention? 
As mentioned in the beginning of this thread, NeighborhoodAttention2D module's forward method only takes one variable as input, which in my case would be one feature map, Is there a way I can use Natten without writing another module / Can I directly use the module defined by @Daniel63656 above?

**Comment by alihassanijr** [comment]
Yes, the modules assume it's a self attention operator, but you can duplicate the module in `src/natten/natten2d.py` and customize it according to your use case.

I'd recommend doing that with the same version of NATTEN you'll end up installing.
All you'd need to do is:
1. Split the QKV linear layer into two, and
2. Modify `forward` and have it take in two inputs, and send the query through the Q projection, and the context through the KV projection.

**Comment by aniket-gupta1** [comment]
@alihassanijr Thanks for all the inputs! I'll try doing that.
